## AutoTokenizer

AI models have different tokenizers for different LLMs. When we want to run them, we dont have to worry about which model uses which tokenizer method. Autotokenizer takes care of that for us.

In [9]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

#### Testing with a sentence

In [10]:
tokenizer("I am going to eat pizza!")

{'input_ids': [101, 1045, 2572, 2183, 2000, 4521, 10733, 999, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

#### We can also pass a list of sentences, as it will have faster processing than processing sentences differenty

In [11]:
raw_inputs = [
    "I've been waiting for a HuggingFace course my whole life.",
    "I hate this so much!",
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

#padding is used to make all  the inputs the same length, this is useful when we want inputs to be of same length

#truncation is there as a safety net. there is a limit on the length of a sentence, and if limit is surpassed, truncation will chop off ends of the sentence.

#return tensort = "pt": this gives the resultant as a pytorch tensor

{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}


Here  

101 - This is called CLS token. Marks the beginning of string  

102 - SEP token. This tells the ending of a sentence.  

Attention Mask - This tells us which of the numbers are real and fake. Because we have given padding to the string, it is filled with 0. We need to tell computer that the numbers are just filler, otherwise it will include the 0's also while the models learning.

### AutoModel

In [12]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModel.from_pretrained(checkpoint)

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertModel LOAD REPORT from: distilbert-base-uncased-finetuned-sst-2-english
Key                   | Status     |  | 
----------------------+------------+--+-
classifier.weight     | UNEXPECTED |  | 
pre_classifier.bias   | UNEXPECTED |  | 
classifier.bias       | UNEXPECTED |  | 
pre_classifier.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


#### Checking with list of Inputs we have.

In [13]:
outputs = model(**inputs)                   # ** here is used to pass the dictionary key-value pairs seperately
print(outputs.last_hidden_state.shape)

torch.Size([2, 16, 768])
